### PR TITLE
[Datasets] Add `read_tfrecords` comment about reading compressed files

### LINKED ISSUE #32205: [Datasets] Support reading compressed TFRecord file
### Description

We found user has use case to read compressed TFRecord file (gzip in the particular case). PyArrow already provide a [pyarrow.CompressedInputStream](https://arrow.apache.org/docs/python/generated/pyarrow.CompressedInputStream.html) wrapper to read compressed file, so we can easily add support to read TFRecord file with common compression algorithms (“bz2”, “brotli”, “gzip”, “lz4” or “zstd”).

### Use case

Unblock users to read TFRecord file with Ray Data for training workload.

### PR DESCRIPTION
**Pull request by scottjlee**
Signed-off-by: Scott Lee <sjl@anyscale.com>

## Why are these changes needed?
It is currently possible to read compressed TFRecord files using `ray.data.read_tfrecords`, but the flow is not made explicitly clear in the function comment. This PR clarifies usage of the function with compressed files and adds an example. 

In a future PR, we will also add a similar example on the [Creating Datasets](https://docs.ray.io/en/latest/data/creating-datasets.html#reading-files-from-storage) user guide page to demonstrate reading compressed files.

## Related issue number
Closes #32205 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/read_api.py [hunk 1]
```diff
@@ -1047,14 +1047,33 @@ def read_tfrecords(
            length  width    species
         0     5.1    3.5  b'setosa'
 
+        We can also read compressed TFRecord files which uses one of the
+        `compression type supported by Arrow <https://arrow.apache.org/docs/python/generated/pyarrow.CompressedInputStream.html>`:
+
+        >>> compressed_path = os.path.join(tempfile.gettempdir(), "data_compressed.tfrecords")
+        >>> options = tf.io.TFRecordOptions(compression_type="ZLIB") # "GZIP" also supported by TensorFlow
```
Comment (c21) [inline python/ray/data/read_api.py:1054]: nit: given we read the gzip file, it's kind of weird that we wrote w/ `zlib` but read w/ `gzip`, can we change it to:

```
>>> options = tf.io.TFRecordOptions(compression_type="GZIP") # "ZLIB" also supported by TensorFlow
```

File: python/ray/data/read_api.py [hunk 2]
```diff
@@ -1047,14 +1047,33 @@ def read_tfrecords(
...
+        >>> with tf.io.TFRecordWriter(path=compressed_path, options=options) as writer:
+        ...     writer.write(example.SerializeToString())
+
+        >>> ds_decompressed = ray.data.read_tfrecords(
```
Comment (c21) [inline python/ray/data/read_api.py:1058]: nit: it's fine to just use `ds`, instead of `ds_decompressed`

File: python/ray/data/read_api.py [hunk 3]
```diff
@@ -1047,14 +1047,33 @@ def read_tfrecords(
...
     Args:
         paths: A single file/directory path or a list of file/directory paths.
             A list of paths can contain both files and directories.
         filesystem: The filesystem implementation to read from.
         parallelism: The requested parallelism of the read. Parallelism may be
             limited by the number of files in the dataset.
         arrow_open_stream_args: Key-word arguments passed to
-            ``pyarrow.fs.FileSystem.open_input_stream``.
+            ``pyarrow.fs.FileSystem.open_input_stream``. To read a compressed TFRecord file,
+            pass the corresponding compression type into ``arrow_open_stream_args``
```
Comment (c21) [inline python/ray/data/read_api.py:1074]: nit:  `into arrow_open_stream_args` can be removed to be more concise.

**Review by c21 — APPROVED** [review]

**Comment by clarkzinzow** [comment]
@scottjlee Doctest failure appears to be related:
```


Document: data/api/doc/ray.data.read_tfrecords
--
  | ----------------------------------------------
  | 2023-02-16 12:53:52,875	WARNING read_api.py:329 -- ⚠️  The number of blocks in this dataset (1) limits its parallelism to 1 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
  | **********************************************************************
  | File "../../python/ray/data/read_api.py", line ?, in default
  | Failed example:
  | ds = ray.data.read_tfrecords(
  | Exception raised:
  | Traceback (most recent call last):
  | File "/opt/miniconda/lib/python3.7/doctest.py", line 1337, in __run
  | compileflags, 1), test.globs)
  | File "/opt/miniconda/lib/python3.7/site-packages/sphinx/ext/doctest.py", line 482, in compile
  | return compile(code, name, self.type, flags, dont_inherit)
  | File "<doctest default[0]>", line 1
  | ds = ray.data.read_tfrecords(
  | ^
  | SyntaxError: unexpected EOF while parsing
  | **********************************************************************
  | 1 items had failures:
  | 1 of  13 in default
  | 13 tests in 1 items.
  | 12 passed and 1 failed.
  | ***Test Failed*** 1 failures.
```

File: python/ray/data/read_api.py [hunk 4]
```diff
@@ -1047,14 +1047,32 @@ def read_tfrecords(
...
+        We can also read compressed TFRecord files which uses one of the
+        `compression type supported by Arrow <https://arrow.apache.org/docs/python/generated/pyarrow.CompressedInputStream.html>`:
+
+        >>> compressed_path = os.path.join(tempfile.gettempdir(), "data_compressed.tfrecords")
+        >>> options = tf.io.TFRecordOptions(compression_type="GZIP") # "ZLIB" also supported by TensorFlow
+        >>> with tf.io.TFRecordWriter(path=compressed_path, options=options) as writer:
+        ...     writer.write(example.SerializeToString())
+
+        >>> ds = ray.data.read_tfrecords(
+                [compressed_path],
```
Comment (clarkzinzow) [inline python/ray/data/read_api.py:1059]: @scottjlee I think you're missing the `...` prefix here under the `>>>` for this line and the following two lines, hence the doctest failure.

File: python/ray/data/read_api.py [hunk 5]
```diff
@@ -1047,14 +1047,32 @@ def read_tfrecords(
            length  width    species
         0     5.1    3.5  b'setosa'
 
+        We can also read compressed TFRecord files which uses one of the
+        `compression type supported by Arrow <https://arrow.apache.org/docs/python/generated/pyarrow.CompressedInputStream.html>`:
```
Comment (c21) [inline python/ray/data/read_api.py:1051]: I think you miss a `_` at the end, see other example [here](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L577).

**Review by clarkzinzow — APPROVED** [review]

**Pull request merged**